tests/api/auth_test.py

```diff
@@ -0,0 +1,59 @@
+import pytest
+
+from zino.api.auth import AuthenticationFailure, authenticate, get_challenge, read_users
+
+
+class TestAuthenticate:
+    def test_when_challenge_response_is_ok_should_return_true(self, secrets_file):
+        good_response = "b3e9ba156949152c5c6f3334cf86333b1daabfe9"
+        assert authenticate(
+            "user1",
+            response=good_response,
+            challenge="a97fbebb3eef2dfcf167645229c0c5fa9e92e3da",
+            secrets_file=secrets_file,
+        )
+
+    def test_when_challenge_response_is_bad_should_raise_error(self, secrets_file):
+        bad_response = "aaa9ba156949152c5c6f3334cf86333b1daabbbb"
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "user1",
+                response=bad_response,
+                challenge="a97fbebb3eef2dfcf167645229c0c5fa9e92e3da",
+                secrets_file=secrets_file,
+            )
+
+    def test_when_user_does_not_exist_should_raise_error(self, secrets_file):
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "nobody",
+                response="doesntmatter",
+                challenge="doesntmatter",
+                secrets_file=secrets_file,
+            )
+
+    def test_when_challenge_is_empty_should_raise_error(self, secrets_file):
+        """Can be removed when challenge-less auth is supported"""
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "user1",
+                response="doesntmatter",
+                secrets_file=secrets_file,
+            )
+
+
+class TestGetChallenge:
+    def test_should_never_return_same_challenge(self):
+        count = 10
+        challenges = set(get_challenge() for _ in range(count))
+        assert len(challenges) == 10, "same challenge was produced more than once"
+
+    def test_should_return_at_least_40_characters(self):
+        assert len(get_challenge()) >= 40
+
+
+def test_read_users_when_given_a_valid_file_then_a_proper_dict_should_be_returned(secrets_file):
+    users = read_users(secrets_file)
+    for user in ("user1", "user2", "user3"):
+        assert user in users
+        assert len(users[user]) == 40
```

Comment:
In one test you check that a challenge is at least 40, and in the next you check that its exactly 40. This is a bit inconsistent, even though it works in this case since you know what the `secret_file` contains. Would be nice to be consistent between these tho for "tests as documentation" purposes, since this kinda implies a requirement for being a "proper" dict is that it contains secrets of exactly length 40.

Feel free to disagree. If this is supposed to test that the secrets in the `secret_file` are returned properly  from `read_users`, then checking the exact length makes some sense. But in that case it might make more sense to check the exact values instead.

The problem here might more be the naming of the test and not the actually testing tho

Reply:
> In one test you check that a challenge is at least 40, and in the next you check that its exactly 40. 

No. In one test, I verify that a *challenge issued by Zino* is at least 40 characters. In the next test, I verify that the *secrets* (no challenges are involved here) parsed from the test data are exactly as long as they were in the test data.

More explicitly, they could test that the secrets are identical to the inputs, as this test would not detect that the code changed the secrets into all `A` characters, for example :grin:

Reply:
Ah, I see. I've misunderstood the connection between challenge and secrets